packages/plugin-pnpm/sources/PnpmLinker.ts

```diff
@@ -0,0 +1,324 @@
+    const customDataKey = getCustomDataKey();
+    const customData = opts.project.installersCustomData.get(customDataKey) as any;
+    if (!customData)
+      throw new UsageError(`The project in ${formatUtils.pretty(opts.project.configuration, `${opts.project.cwd}/package.json`, formatUtils.Type.PATH)} doesn't seem to have been installed - running an install there might help`);
+
+    const nmRootLocation = location.match(/(^.*\/node_modules\/(@[^/]*\/)?[^/]+)(\/.*$)/);
+    if (nmRootLocation) {
+      const nmLocator = customData.locatorByPath.get(nmRootLocation[1]);
+      if (nmLocator) {
+        return nmLocator;
+      }
+    }
+
+    let nextPath = location;
+    let currentPath = location;
+    do {
+      currentPath = nextPath;
+      nextPath = ppath.dirname(currentPath);
+
+      const locator = customData.locatorByPath.get(currentPath);
+      if (locator) {
+        return locator;
+      }
+    } while (nextPath !== currentPath);
+
+    return null;
+  }
+
+  makeInstaller(opts: LinkOptions) {
+    return new PnpmInstaller(opts);
+  }
+}
+
+class PnpmInstaller implements Installer {
+  private asyncActions = new AsyncActions();
+  private packageLocations = new Map<LocatorHash, PortablePath>();
+
+  constructor(private opts: LinkOptions) {
+    // Nothing to do
+  }
+
+  getInstallerSettings() {
+    return {
+      installPackageConcurrency: 10,
+    };
+  }
+
+  getCustomDataKey() {
+    return getCustomDataKey();
+  }
+
+  private customData: PnpmCustomData = {
+    locatorByPath: new Map(),
+  };
+
+  attachCustomData(customData: any) {
+    this.customData = customData;
+  }
+
+  async installPackage(pkg: Package, fetchResult: FetchResult) {
+    switch (pkg.linkType) {
+      case LinkType.SOFT: return this.installPackageSoft(pkg, fetchResult);
+      case LinkType.HARD: return this.installPackageHard(pkg, fetchResult);
+    }
+
+    throw new Error(`Assertion failed: Unsupported package link type`);
+  }
+
+  async installPackageSoft(pkg: Package, fetchResult: FetchResult) {
+    const pkgPath = ppath.resolve(fetchResult.packageFs.getRealPath(), fetchResult.prefixPath);
+    this.packageLocations.set(pkg.locatorHash, pkgPath);
+
+    return {
+      packageLocation: pkgPath,
+      buildDirective: null,
+    };
+  }
+
+  async installPackageHard(pkg: Package, fetchResult: FetchResult) {
+    const pkgPath = getPackageLocation(pkg, {project: this.opts.project});
+
+    this.customData.locatorByPath.set(pkgPath, structUtils.stringifyLocator(pkg));
+    this.packageLocations.set(pkg.locatorHash, pkgPath);
+
+    this.asyncActions.set(pkg.locatorHash, async () => {
+      await xfs.mkdirPromise(pkgPath, {recursive: true});
+
+      // Copy the package source into the <root>/n_m/.store/<hash> directory, so
+      // that we can then create symbolic links to it later.
+      await xfs.copyPromise(pkgPath, fetchResult.prefixPath, {
+        baseFs: fetchResult.packageFs,
```

Comment:
Wont the core release this before the copy completes?

Reply:
Of course 🤦‍♀️ completely forgot that. I've implemented a `holdFetchResult` that instructs the core not to reclaim the `fetchResult` until the given promise has resolved. Does that sound more elegant than the previous design?

Reply:
It's better but I had more of a ref counter / `shared_ptr` like system in mind so the `fetchResult` itself contains the logic and not the core

Reply:
This would likely require a heavier refactoring than what I'm confident in doing for this feature at the moment 🤔

Reply:
That's fair, I was planning on changing the core to pass in a FS wrapping `ZipOpenFS` which has this logic implemented already but that's for another time